include/net/mqtt.h

```diff
@@ -1,474 +1,685 @@
+	 */
+	const void *broker;
+
+	/** User name (if any) to be used for the connection. NULL indicates
+	 *  no user name.
+	 */
+	struct mqtt_utf8 *user_name;
+
+	/** Password (if any) to be used for the connection. Note that if
+	 *  password is provided, user name shall also be provided. NULL
+	 *  indicates no password.
+	 */
+	struct mqtt_utf8 *password;
+
+	/** Will topic and QoS. Can be NULL. */
+	struct mqtt_topic *will_topic;
+
+	/** Will message. Can be NULL. Non NULL value valid only if will topic
+	 *  is not NULL.
+	 */
+	struct mqtt_utf8 *will_message;
+
+	/** Application callback registered with the module to get MQTT events.
+	 */
+	mqtt_evt_cb_t evt_cb;
+
+	/** Receive buffer used for MQTT packet reception in RX path. */
+	u8_t *rx_buf;
+
+	/** Size of receive buffer. */
+	u32_t rx_buf_size;
+
+	/** Transmit buffer used for creating MQTT packet in TX path. */
+	u8_t *tx_buf;
+
+	/** Size of transmit buffer. */
+	u32_t tx_buf_size;
+
+	/** MQTT protocol version. */
+	u8_t protocol_version;
+
+	/** Will retain flag, 1 if will message shall be retained persistently.
+	 */
+	u8_t will_retain : 1;
+
+	/** Clean session flag indicating a fresh (1) or a retained session (0).
+	 *  Default is 1.
+	 */
+	u8_t clean_session : 1;
+};
 
 /**
- * Sends the MQTT PUBCOMP message with the given packet id
+ * @brief Initializes the client instance.
  *
- * @param [in] ctx MQTT context structure
- * @param [in] id MQTT Packet Identifier
+ * @param[in] client Client instance for which the procedure is requested.
+ *                   Shall not be NULL.
  *
- * @retval 0 on success
- * @retval -EINVAL
- * @retval -ENOMEM
- * @retval -EIO
+ * @note Shall be called before connecting the client in order to avoid
+ *       unexpected behavior caused by uninitialized parameters.
  */
-int mqtt_tx_pubcomp(struct mqtt_ctx *ctx, u16_t id);
+void mqtt_client_init(struct mqtt_client *client);
 
 /**
- * Sends the MQTT PUBREC message with the given packet id
- *
- * @param [in] ctx MQTT context structure
- * @param [in] id MQTT Packet Identifier
- * @retval 0 on success
- * @retval -EINVAL
- * @retval -ENOMEM
- * @retval -EIO
+ * @brief API to request new MQTT client connection.
+ *
+ * @param[in] client Client instance for which the procedure is requested.
+ *                   Shall not be NULL.
+ *
+ * @note This memory is assumed to be resident until mqtt_disconnect is called.
+ * @note Any subsequent changes to parameters like broker address, user name,
+ *       device id, etc. have no effect once MQTT connection is established.
+ *
+ * @return 0 or a negative error code (errno.h) indicating reason of failure.
+ *
+ * @note Default protocol revision used for connection request is 3.1.1. Please
+ *       set client.protocol_version = MQTT_VERSION_3_1_0 to use protocol 3.1.0.
+ * @note If more than one simultaneous client connections are needed, please
+ *       modify :option:`CONFIG_MQTT_MAX_CLIENTS` to override default of 1.
```

Comment:
A consequence of the good design - one less option for users to care about! ;-) (Please remove.)